indexer/services/ender/__tests__/handlers/stateful-order/stateful-order-replacement-handler.test.ts

```diff
@@ -0,0 +1,234 @@
+      clientId: defaultNewOrder.orderId!.clientId.toString(),
+      clobPairId: defaultNewOrder.orderId!.clobPairId.toString(),
+      side: OrderSide.BUY,
+      size: getSize(defaultNewOrder, testConstants.defaultPerpetualMarket),
+      totalFilled: '0',
+      price: getPrice(defaultNewOrder, testConstants.defaultPerpetualMarket),
+      type: OrderType.LIMIT, // TODO: Add additional order types once we support
+      status: OrderStatus.OPEN,
+      timeInForce: protocolTranslations.protocolOrderTIFToTIF(defaultNewOrder.timeInForce),
+      reduceOnly: defaultNewOrder.reduceOnly,
+      orderFlags: defaultNewOrder.orderId!.orderFlags.toString(),
+      goodTilBlock: null,
+      goodTilBlockTime: protocolTranslations.getGoodTilBlockTime(defaultNewOrder),
+      createdAtHeight: '3',
+      clientMetadata: '0',
+      triggerPrice: null,
+      updatedAt: defaultDateTime.toISO(),
+      updatedAtHeight: defaultHeight.toString(),
+    });
+
+    const expectedOffchainUpdate: OffChainUpdateV1 = {
+      orderReplace: {
+        oldOrderId: defaultOldOrder.orderId!,
+        order: defaultNewOrder,
+        placementStatus: OrderPlaceV1_OrderPlacementStatus.ORDER_PLACEMENT_STATUS_OPENED,
+      },
+    };
+    expectVulcanKafkaMessage({
+      producerSendMock,
+      orderId: defaultNewOrder.orderId!,
+      offchainUpdate: expectedOffchainUpdate,
+      headers: { message_received_timestamp: kafkaMessage.timestamp, event_type: 'StatefulOrderReplacement' },
+    });
+    if (emitSubaccountMessage) {
+      expectOrderSubaccountKafkaMessage(
+        producerSendMock,
+        defaultNewOrder.orderId!.subaccountId!,
+        newOrder!,
+        defaultHeight.toString(),
+        transactionIndex,
+      );
+    }
+  });
+
+  it.each([
+    ['stateful order replacement', defaultStatefulOrderReplacementEvent],
+  ])('successfully upserts order with %s', async (
+    _name: string,
+    statefulOrderReplacementEvent: StatefulOrderEventV1,
+  ) => {
+    const subaccountId: string = SubaccountTable.subaccountIdToUuid(
+      defaultNewOrder.orderId!.subaccountId!,
+    );
+    const clientId: string = defaultNewOrder.orderId!.clientId.toString();
+    const clobPairId: string = defaultNewOrder.orderId!.clobPairId.toString();
+    await OrderTable.create({
+      subaccountId,
+      clientId,
+      clobPairId,
+      side: OrderSide.SELL,
+      size: '100',
+      totalFilled: '0',
+      price: '200',
+      type: OrderType.LIMIT,
+      status: OrderStatus.CANCELED,
+      timeInForce: TimeInForce.GTT,
+      reduceOnly: true,
+      orderFlags: '0',
+      goodTilBlockTime: protocolTranslations.getGoodTilBlockTime(defaultNewOrder),
+      createdAtHeight: '1',
+      clientMetadata: '0',
+      updatedAt: defaultDateTime.toISO(),
+      updatedAtHeight: '0',
+    });
+    const kafkaMessage: KafkaMessage = createKafkaMessageFromStatefulOrderEvent(
+      statefulOrderReplacementEvent,
+    );
+
+    await onMessage(kafkaMessage);
```

Comment:
Does the old order not exist in the db, and so we aren't checking it in this test?
Shouldn't the old order always exist if an StatefulOrderReplacement event is sent? Do we need error handling for this case?

Reply:
I copied this test from `stateful-order-placement-handler.test.ts` and it's a behavior that stateful order replacements should maintain. We check the old order in the test above. I think it'd be fine for order replacements to succeed at placing the new order if the old order doesn't exist and to handle that gracefully. What do you think?

Reply:
hmmm in what case would this happen? Seems to me like a replacement should replace an order. If it's a new order, we should handle via OrderPlacement?

Reply:
changed the test to test for when the old and new order have the same order ID